localstack/aws/protocol/router.py

```diff
@@ -0,0 +1,234 @@
+        input_shape = op.input_shape
+        if isinstance(input_shape, StructureShape):
+            for required_member in input_shape.required_members:
+                member_shape = input_shape.members[required_member]
+                location = member_shape.serialization.get("location")
+                if location is not None:
+                    if location == "header":
+                        header_name = member_shape.serialization.get("name")
+                        header_args.append(header_name)
+                    elif location == "querystring":
+                        query_name = member_shape.serialization.get("name")
+                        # do not overwrite potentially already existing query params with specific values
+                        if query_name not in query_args:
+                            # an empty list defines a required query param only needs to be present
+                            # (no specific value will be enforced when matching)
+                            query_args[query_name] = []
+
+        return _HttpOperation(op, path, method, query_args, header_args)
+
+
+class _RequestMatchingRule(Rule):
+    """
+    A Werkzeug Rule extension which initially acts as a normal rule (i.e. matches a path and method).
+
+    This rule matches if one of its sub-rules _might_ match.
+    It cannot be assumed that one of the fine-grained rules matches, just because this rule initially matches.
+    If this rule matches, the caller _must_ call `match_request` in order to find the actual fine-grained matching rule.
+    The result of `match_request` is only meaningful if this wrapping rule also matches.
+    """
+
+    def __init__(self, string: str, *args, operations: List[_HttpOperation], **kwargs) -> None:
+        super().__init__(string, *args, **kwargs)
+        # Create a rule which checks all required arguments (not only the path and method)
+        rules = [
+            _RequiredArgsRule(op.operation, op.query_args, op.header_args) for op in operations
+        ]
+        # Sort the rules descending based on their rule score
+        # (i.e. the first matching rule will have the highest score)=
+        self.rules = sorted(rules, key=lambda rule: rule.match_score, reverse=True)
+
+    def match_request(self, request: Request) -> _RequiredArgsRule:
+        """
+        Function which needs to be called by a caller if the _RequestMatchingRule already matched using Werkzeug's
+        default matching mechanism.
+
+        :param request: to perform the fine-grained matching on
+        :return: matching fine-grained rule
+        :raises: NotFound if none of the fine-grained rules matches
+        """
+        try:
+            # return the first matching rule or raise a NotFound error if no rule matches
+            return next(
+                filter(lambda rule: rule.matches(request.args, request.headers), self.rules)
+            )
+        except StopIteration:
+            raise NotFound()
```

Comment:
FWIW, i've seen this pattern several times now and I have to say I'm not super fond of it. Isn't this version a lot easier to read?:
```python
for rule in self.rules:
    if rule.matches(request.args, request.headers):
        return rule
raise NotFound()
```

plus it avoids the extra `StopIteration` except overhead.
is there any benefit to the `next`,`filter`,`lambda` nesting i may be missing?

Reply:
I guess it's question of taste. I like the functional approach a bit more, but I see your point (and mine is highly subjective). I'll change it.